Format numbers in larq model summaries to be more readable. Closes #192.

AdamHillier opened the pull request:
An example approach that formats numbers with three-significant figures and a magnitude (e.g. `k`, `M`, `B`, et cetera). This works for all numbers up to ~999 trillion. For larger numbers, scientific notation is used.

Illustrative examples:

```
>>> for i in range(1, 20):
...     num = int("1234567890123456789"[:i])
...     print("Number:", num, " " * (20 - i), "Formatted:", _number_as_readable_str(num))
...
Number: 1                     Formatted: 1
Number: 12                    Formatted: 12
Number: 123                   Formatted: 123
Number: 1234                  Formatted: 1.23 k
Number: 12345                 Formatted: 12.3 k
Number: 123456                Formatted: 123 k
Number: 1234567               Formatted: 1.23 M
Number: 12345678              Formatted: 12.3 M
Number: 123456789             Formatted: 123 M
Number: 1234567890            Formatted: 1.23 B
Number: 12345678901           Formatted: 12.3 B
Number: 123456789012          Formatted: 123 B
Number: 1234567890123         Formatted: 1.23 T
Number: 12345678901234        Formatted: 12.3 T
Number: 123456789012345       Formatted: 123 T
Number: 1234567890123456      Formatted: 1.23E+15
Number: 12345678901234567     Formatted: 1.23E+16
Number: 123456789012345678    Formatted: 1.23E+17
Number: 1234567890123456789   Formatted: 1.23E+18

>>> for i in range(1, 20):
...     num = int("99999999999999999999"[:i])
...     print("Number:", num, " " * (20 - i), "Formatted:", _number_as_readable_str(num))
...
Number: 9                     Formatted: 9
Number: 99                    Formatted: 99
Number: 999                   Formatted: 999
Number: 9999                  Formatted: 10.0 k
Number: 99999                 Formatted: 100 k
Number: 999999                Formatted: 1.00 M
Number: 9999999               Formatted: 10.0 M
Number: 99999999              Formatted: 100 M
Number: 999999999             Formatted: 1.00 B
Number: 9999999999            Formatted: 10.0 B
Number: 99999999999           Formatted: 100 B
Number: 999999999999          Formatted: 1.00 T
Number: 9999999999999         Formatted: 10.0 T
Number: 99999999999999        Formatted: 100 T
Number: 999999999999999       Formatted: 1.00E+15
Number: 9999999999999999      Formatted: 1.00E+16
Number: 99999999999999999     Formatted: 1.00E+17
Number: 999999999999999999    Formatted: 1.00E+18
Number: 9999999999999999999   Formatted: 1.00E+19
```

lgeiger commented:
That definitely looks a lot better :+1:

Though in some cases I think it still might be useful to know the exact number of parameters (e.g. when reimplementing a model from scratch as a sanity check).
What do you think about just using spaces to separate the numbers for better readability e.g `999 999 999`?
@koenhelwegen Do you have an opinion on that?

koenhelwegen commented:
Nice! I think this will almost always be preferable, but we could add a flag in case someone requires the precise numbers?

lgeiger commented:
> I think this will almost always be preferable, but we could add a flag in case someone requires the precise numbers?

Thinking about this again, I agree. I don't really want to introduce another flag anyway, and if we really need precise numbers we can still use the built in Keras model summary.

AdamHillier commented:
I've updated it to use three significant figures and added a flag so that the precise values can be used instead if wanted, but by default the readable totals are used. I added an extra test to cover the non-default -- the snapshot shows no difference with the previous behaviour when readable totals are disabled.

AdamHillier commented:
And I've edited my original comment to update the example formatted numbers.

AdamHillier commented:
After discussion with @lgeiger I've removed the flag - if necessary we can add it back in at a later date.

lgeiger commented on larq/models.py:
```diff
@@ -48,6 +48,34 @@ def _bitsize_as_str(bitsize):
+def _number_as_readable_str(num):
+    # The initial rounding here is necessary so that e.g. `999000` gets
+    # formatted as `1.000 M` rather than `1000 k`
+    num = float("{:.3g}".format(num))
```
```suggestion
    num = float(f"{num:.3g}")
```

lgeiger commented on larq/models.py:
```diff
@@ -48,6 +48,34 @@ def _bitsize_as_str(bitsize):
+    # For numbers that are at least 1000 trillion (1 quadrillion) format with
+    # scientific notation (3 s.f. = 2 d.p. in scientific notation).
+    if num >= 1e15:
+        return "{:#.2E}".format(num)
```
```suggestion
        return f"{num:#.2E}"
```

lgeiger commented on larq/models.py:
```diff
@@ -48,6 +48,34 @@ def _bitsize_as_str(bitsize):
+
+    # ':#.3g' formats the number with 3 significant figures, without stripping
+    # trailing zeros.
+    num = "{:#.3g}".format(num).rstrip(".")
```
```suggestion
    num = f"{num:#.3g}".rstrip(".")
```

AdamHillier commented:
> What do you think about using f-strings instead of the .format syntax?

Yep that looks nicer, I've updated.
